ice-phishing/src/agent.js

```diff
@@ -0,0 +1,256 @@
+  const approvalEvents = [
+    ...txEvent.filterLog(approvalEventSigErc20),
+    ...txEvent.filterLog(approvalEventSigErc721),
+    ...txEvent.filterLog(approvalForAllEventSig),
+  ];
+
+  const transferEvents = [
+    ...txEvent.filterLog(transferEventSigErc20),
+    ...txEvent.filterLog(transferEventSigErc721),
+    ...txEvent.filterLog(erc1155transferEventSigs),
+  ];
+
+  // Don't process transfer events if there are no approvals
+  // Both ERC20 and ERC721 emit Approve on transferFrom
+  if (approvalEvents.length === 0) return findings;
+
+  await Promise.all(approvalEvents.map(async (event) => {
+    const { address: asset, name } = event;
+    const {
+      owner,
+      spender,
+      value,
+      tokenId,
+      approved,
+    } = event.args;
+
+    const isApprovalForAll = (name === 'ApprovalForAll');
+
+    // Filter out approval revokes
+    if (isApprovalForAll && !approved) return;
+    if (value?.eq(0)) return;
+    if (spender === ADDRESS_ZERO) return;
+
+    // When transfering ERC20 tokens an Approval event is emitted with lower value
+    // We should ignore these Approval events because they are duplicates
+    const isAlreadyApproved = (tokenId)
+      ? false
+      : approvals[spender]?.some((a) => a.owner === owner);
+
+    if (isAlreadyApproved) return;
+
+    // Skip if the owner or is not EOA
+    const ownerType = await getAddressType(owner, cachedAddresses, blockNumber, chainId, true);
+    if (ownerType === AddressType.UnverifiedContract
+      || ownerType === AddressType.VerifiedContract
+    ) return;
+
+    // Skip if the spender
+    // has high nonce (probably CEX)
+    // is verified contract
+    // or is ignored address
+    const spenderType = await getAddressType(spender, cachedAddresses, blockNumber, chainId, false);
+    if (
+      spenderType === AddressType.EoaWithHighNonce
+      || spenderType === AddressType.VerifiedContract
+      || spenderType.startsWith('Ignored')) return;
+
+    // Initialize the approvals array for the spender if it doesn't exist
+    if (!approvals[spender]) approvals[spender] = [];
+
+    console.log('Detected possible malicious approval');
+    console.log(`owner: ${owner}`);
+    console.log(`spender: ${spender}`);
+    console.log(`asset: ${asset}`);
+
+    // Update the approvals for the spender
+    approvals[spender].push({
+      asset,
+      owner,
+      hash,
+      timestamp,
+      tokenId,
+      isApprovalForAll,
+    });
+
+    // Filter out old approvals
+    approvals[spender] = approvals[spender].filter((a) => timestamp - a.timestamp < TIME_PERIOD);
+
+    // Ignore the address until the end of the period if there are a lot of approvals
+    if (approvals[spender].length > maxAddressAlertsPerPeriod) {
+      const newType = (spenderType === AddressType.EoaWithLowNonce)
+        ? AddressType.IgnoredEoa
+        : AddressType.IgnoredContract;
+      cachedAddresses.set(spender, newType);
+    }
+
+    if (approvals[spender].length > actionThreshold) {
```

Comment:
i recommend making this two different values; one for approve count threshold and one for transfer count threshold